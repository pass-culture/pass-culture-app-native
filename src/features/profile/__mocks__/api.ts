export enum CHANGE_EMAIL_ERROR_CODE {
  TOKEN_EXISTS = 'TOKEN_EXISTS',
  INVALID_PASSWORD = 'INVALID_PASSWORD',
  EMAIL_UPDATE_ATTEMPTS_LIMIT = 'EMAIL_UPDATE_ATTEMPTS_LIMIT',
}
