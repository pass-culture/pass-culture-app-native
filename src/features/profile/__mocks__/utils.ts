export const isUserExBeneficiary = jest.fn().mockReturnValue(false)
export const isUserBeneficiary = jest.fn().mockReturnValue(true)
export const useIsUserUnderageBeneficiary = jest.fn().mockReturnValue(false)
export const isUserUnderageBeneficiary = jest.fn().mockReturnValue(false)
export const useIsUserUnderage = jest.fn().mockReturnValue(false)
export const matchSubscriptionMessageIconToSvg = jest
  .fn()
  .mockReturnValue('matchSubscriptionMessageIconToSvg')

export const handleCallToActionLink = jest.fn()

export const computeCredit = jest.fn().mockReturnValue(3)
