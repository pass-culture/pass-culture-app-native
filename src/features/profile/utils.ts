/**
 * Formats an iso date to a slashed french date.
 * @param ISOBirthday the birthday date into the ISO 8601 format %Y-%m-%dT%H:%M:%S
 */
export function computeEligibilityExpiracy(ISOBirthday: string) {
  const date = new Date(ISOBirthday)
  date.setFullYear(date.getFullYear() + 19)
  date.setDate(date.getDate() - 1)
  date.setHours(23)
  date.setMinutes(59)
  date.setSeconds(59)
  return date
}
