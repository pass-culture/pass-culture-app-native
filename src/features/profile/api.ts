import { useMutation, useQueryClient } from 'react-query'

import { api } from 'api/api'
import { UserProfileResponse, UserProfileUpdateRequest } from 'api/gen'
import { QueryKeys } from 'libs/queryKeys'

export enum CHANGE_EMAIL_ERROR_CODE {
  TOKEN_EXISTS = 'TOKEN_EXISTS',
  INVALID_PASSWORD = 'INVALID_PASSWORD',
  EMAIL_UPDATE_ATTEMPTS_LIMIT = 'EMAIL_UPDATE_ATTEMPTS_LIMIT',
}

export interface ChangeEmailRequest {
  email: string
  password: string
}

type ResetRecreditAmountToShowMutationOptions = {
  onSuccess: () => void
  onError: (error: unknown) => void
}

export function useUpdateProfileMutation(
  onSuccessCallback: (data: UserProfileResponse) => void,
  onErrorCallback: (error: unknown) => void
) {
  const client = useQueryClient()
  return useMutation((body: UserProfileUpdateRequest) => api.postnativev1profile(body), {
    onSuccess(response: UserProfileResponse) {
      client.setQueryData(QueryKeys.USER_PROFILE, (old: UserProfileResponse | undefined) => ({
        ...(old || {}),
        ...response,
      }))
      onSuccessCallback(response)
    },
    onError: onErrorCallback,
  })
}

export function useResetRecreditAmountToShow({
  onSuccess,
  onError,
}: ResetRecreditAmountToShowMutationOptions) {
  return useMutation(() => api.postnativev1resetRecreditAmountToShow(), {
    onSuccess,
    onError,
  })
}
