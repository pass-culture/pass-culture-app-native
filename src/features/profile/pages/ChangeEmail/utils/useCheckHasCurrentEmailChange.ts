import { useQuery } from 'react-query'

import { api } from 'api/api'
import { QueryKeys } from 'libs/queryKeys'

export const useCheckHasCurrentEmailChange = () => {
  const { data: currentEmailChangeTimestampResponse } = useQuery(
    QueryKeys.EMAIL_CHANGE_EXPIRATION_TIMESTAMP,
    () => api.getnativev1profiletokenExpiration()
  )
  const hasCurrentEmailChange = !!currentEmailChangeTimestampResponse?.expiration
  return { hasCurrentEmailChange }
}
