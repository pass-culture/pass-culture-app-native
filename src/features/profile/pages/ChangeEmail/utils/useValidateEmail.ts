import { t } from '@lingui/macro'

import { useUserProfileInfo } from 'features/home/api'
import { isEmailValid } from 'ui/components/inputs/emailCheck'

export const useIsCurrentUserEmail = (email: string): boolean => {
  const { data: user } = useUserProfileInfo()
  return email.toLowerCase() === user?.email?.toLowerCase()
}

export function useValidateEmail(email: string): useValidateEmailReturn {
  const isCurrentUserEmail = useIsCurrentUserEmail(email)

  if (email.length === 0) return { emailErrorMessage: null, isEmailValid: false }
  if (!isEmailValid(email))
    return {
      emailErrorMessage: t`L'e-mail renseigné est incorrect. Exemple de format attendu\u00a0: edith.piaf@email.fr`,
      isEmailValid: false,
    }
  if (isCurrentUserEmail)
    return {
      emailErrorMessage: t`L'e-mail saisi est identique à votre e-mail actuel`,
      isEmailValid: false,
    }
  return { emailErrorMessage: null, isEmailValid: true }
}

type useValidateEmailReturn = { emailErrorMessage: string | null; isEmailValid: boolean }
