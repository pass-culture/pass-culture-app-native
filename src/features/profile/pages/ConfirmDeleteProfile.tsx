import { t } from '@lingui/macro'
import { useNavigation } from '@react-navigation/native'
import React from 'react'
import styled from 'styled-components/native'

import { useAccountSuspend } from 'features/auth/api'
import { useLogoutRoutine } from 'features/auth/AuthContext'
import { UseNavigationType } from 'features/navigation/RootNavigator'
import { AppButton } from 'ui/components/buttons/AppButton'
import { SNACK_BAR_TIME_OUT, useSnackBarContext } from 'ui/components/snackBar/SnackBarContext'
import { Background } from 'ui/svg/Background'
import { Error } from 'ui/svg/icons/Error'
import { ColorsEnum, getSpacing, Spacer, Typo } from 'ui/theme'

export function ConfirmDeleteProfile() {
  const { navigate, goBack } = useNavigation<UseNavigationType>()
  const signOut = useLogoutRoutine()
  const { showErrorSnackBar } = useSnackBarContext()

  function onAccountSuspendSuccess() {
    navigate('DeleteProfileSuccess')
    signOut()
  }

  function onAccountSuspendFailure() {
    showErrorSnackBar({
      message: t`Une erreur s’est produite pendant le chargement.`,
      timeout: SNACK_BAR_TIME_OUT,
    })
  }

  const { mutate: notifyAccountSuspend, isLoading } = useAccountSuspend(
    onAccountSuspendSuccess,
    onAccountSuspendFailure
  )

  return (
    <Container>
      <Background />
      <Spacer.TopScreen />
      <Spacer.Flex />
      <Error />

      <Spacer.Column numberOfSpaces={6} />

      <CenteredContainer>
        <CenteredText>
          <Typo.Title2 color={ColorsEnum.WHITE}>
            {t`Es-tu sûr de vouloir supprimer ton compte ?`}
          </Typo.Title2>
        </CenteredText>

        <Spacer.Column numberOfSpaces={4} />

        <CenteredText>
          <Typo.Body color={ColorsEnum.WHITE}>
            {t`Cela entraînera l'annulation de l'ensemble de tes réservations en cours, ainsi que la suppression définitive de ton crédit pass Culture si tu en bénéficies.`}
          </Typo.Body>
        </CenteredText>
      </CenteredContainer>

      <Spacer.Flex />

      <Row>
        <ButtonContainer>
          <AppButton
            title={t`Supprimer mon compte`}
            onPress={notifyAccountSuspend}
            backgroundColor={ColorsEnum.WHITE}
            textColor={ColorsEnum.PRIMARY}
            loadingIconColor={ColorsEnum.PRIMARY}
            buttonHeight="small"
            isLoading={isLoading}
          />
          <Spacer.Column numberOfSpaces={2} />
          <AppButton
            title={t`Abandonner`}
            onPress={goBack}
            backgroundColor={ColorsEnum.TRANSPARENT}
            textColor={ColorsEnum.WHITE}
            loadingIconColor={ColorsEnum.WHITE}
            buttonHeight="tall"
          />
        </ButtonContainer>
      </Row>
      <Spacer.BottomScreen />
    </Container>
  )
}

const Container = styled.View({
  flex: 1,
  alignItems: 'center',
})

const Row = styled.View({ flexDirection: 'row' })

const CenteredContainer = styled.View({
  alignItems: 'center',
  justifyContent: 'center',
  textAlign: 'center',
  marginHorizontal: getSpacing(6),
})

const ButtonContainer = styled.View({
  flex: 1,
  marginHorizontal: getSpacing(6),
  paddingBottom: getSpacing(6),
})

const CenteredText = styled.Text({
  textAlign: 'center',
})
