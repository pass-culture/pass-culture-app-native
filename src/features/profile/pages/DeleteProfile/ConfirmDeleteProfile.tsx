import { t } from '@lingui/macro'
import { useNavigation } from '@react-navigation/native'
import React from 'react'
import styled from 'styled-components/native'

import { useAccountSuspend } from 'features/auth/api'
import { useLogoutRoutine } from 'features/auth/AuthContext'
import { UseNavigationType } from 'features/navigation/RootNavigator'
import { getTabNavConfig } from 'features/navigation/TabBar/helpers'
import { useGoBack } from 'features/navigation/useGoBack'
import { ButtonPrimaryWhite } from 'ui/components/buttons/ButtonPrimaryWhite'
import { ButtonTertiaryWhite } from 'ui/components/buttons/ButtonTertiaryWhite'
import { GenericInfoPage } from 'ui/components/GenericInfoPage'
import { SNACK_BAR_TIME_OUT, useSnackBarContext } from 'ui/components/snackBar/SnackBarContext'
import { ErrorIllustration } from 'ui/svg/icons/ErrorIllustration'
import { Typo } from 'ui/theme'

export function ConfirmDeleteProfile() {
  const { navigate } = useNavigation<UseNavigationType>()
  const { goBack } = useGoBack(...getTabNavConfig('Profile'))

  const signOut = useLogoutRoutine()
  const { showErrorSnackBar } = useSnackBarContext()

  function onAccountSuspendSuccess() {
    navigate('DeleteProfileSuccess')
    signOut()
  }

  function onAccountSuspendFailure() {
    showErrorSnackBar({
      message: t`Une erreur s’est produite pendant le chargement.`,
      timeout: SNACK_BAR_TIME_OUT,
    })
  }

  const { mutate: notifyAccountSuspend, isLoading } = useAccountSuspend(
    onAccountSuspendSuccess,
    onAccountSuspendFailure
  )

  return (
    <GenericInfoPage
      title={t`Es-tu sûr de vouloir supprimer ton compte\u00a0?`}
      icon={ErrorIllustration}
      buttons={[
        <ButtonPrimaryWhite
          key={1}
          isLoading={isLoading}
          wording={t`Supprimer mon compte`}
          onPress={notifyAccountSuspend}
        />,
        <ButtonTertiaryWhite key={2} wording={t`Abandonner`} onPress={goBack} />,
      ]}>
      <StyledBody>{t`Cela entraînera l'annulation de l'ensemble de tes réservations en cours, ainsi que la suppression définitive de ton crédit pass Culture si tu en bénéficies.`}</StyledBody>
    </GenericInfoPage>
  )
}

const StyledBody = styled(Typo.Body)(({ theme }) => ({
  color: theme.colors.white,
  textAlign: 'center',
}))
