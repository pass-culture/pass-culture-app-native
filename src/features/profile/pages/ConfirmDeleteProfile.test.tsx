import React from 'react'
import { UseQueryResult } from 'react-query'

import { navigate } from '__mocks__/@react-navigation/native'
import { goBack } from '__mocks__/@react-navigation/native'
import { UserProfileResponse } from 'api/gen'
import { contactSupport } from 'features/auth/support.services'
import { fireEvent, render } from 'tests/utils'

import { ConfirmDeleteProfile } from './ConfirmDeleteProfile'

jest.mock('features/home/api', () => ({
  useUserProfileInfo: jest.fn(
    () =>
      ({
        isLoading: false,
        data: { email: 'email2@domain.ext', firstName: 'Jean', isBeneficiary: false },
      } as UseQueryResult<UserProfileResponse>)
  ),
}))

describe('ConfirmDeleteProfile component', () => {
  it('should render confirm delete profile', () => {
    const renderAPI = render(<ConfirmDeleteProfile />)
    expect(renderAPI.toJSON()).toMatchSnapshot()
  })

  it('should open mail app when clicking on "Je supprime mon compte" button', () => {
    const renderAPI = render(<ConfirmDeleteProfile />)
    fireEvent.press(renderAPI.getByText(`Je supprime mon compte`))
    expect(contactSupport.forAccountDeletion).toBeCalledWith('email2@domain.ext')
  })

  it('should redirect to LegalNotices when clicking on "Abandonner" button', () => {
    const renderAPI = render(<ConfirmDeleteProfile />)
    fireEvent.press(renderAPI.getByText(`Abandonner`))
    expect(goBack).toBeCalled()
    expect(navigate).toBeCalledWith('LegalNotices')
  })
})
