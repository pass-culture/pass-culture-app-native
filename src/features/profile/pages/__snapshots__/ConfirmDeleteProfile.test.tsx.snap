// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDeleteProfile component should render confirm delete profile 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "height": "100%",
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        },
      ]
    }
  >
    <View
      height="100%"
      width="100%"
    >
      <Text>
        undefined-SVG-Mock
      </Text>
    </View>
  </View>
  <View
    customHeight={16}
    style={
      Array [
        Object {
          "height": 16,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  />
  <View
    height={75}
    width={116}
  >
    <Text>
      undefined-SVG-Mock
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexGrow": 1,
          "justifyContent": "center",
          "marginHorizontal": 32,
          "textAlign": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      <Text
        color="#ffffff"
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontFamily": "Montserrat-MediumItalic",
              "fontSize": 24,
              "lineHeight": 28,
            },
          ]
        }
      >
        Es-tu sûr de vouloir supprimer ton compte ?
      </Text>
    </Text>
    <View
      numberOfSpaces={4}
      style={
        Array [
          Object {
            "height": 16,
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      <Text
        color="#ffffff"
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontFamily": "Montserrat-Regular",
              "fontSize": 15,
              "lineHeight": 20,
            },
          ]
        }
      >
        Tu as 30 jours pour te rétracter en envoyant un message à support@passculture.app
      </Text>
    </Text>
    <View
      numberOfSpaces={4}
      style={
        Array [
          Object {
            "height": 16,
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      <Text
        color="#ffffff"
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontFamily": "Montserrat-Regular",
              "fontSize": 15,
              "lineHeight": 20,
            },
          ]
        }
      >
        Une fois ce délai écoulé, tes données seront supprimées de nos fichier
      </Text>
    </Text>
  </View>
  <View
    numberOfSpaces={8}
    style={
      Array [
        Object {
          "height": 32,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "marginHorizontal": 32,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 24,
            "borderWidth": 0,
            "flexDirection": "row",
            "height": 48,
            "justifyContent": "center",
            "maxWidth": 500,
            "opacity": 1,
            "paddingBottom": 2,
            "paddingLeft": 2,
            "paddingRight": 2,
            "paddingTop": 2,
            "width": "100%",
          }
        }
        testID="button-container"
      >
        <Text
          adjustsFontSizeToFit={false}
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#eb0055",
                "fontFamily": "Montserrat-Bold",
                "fontSize": 15,
                "lineHeight": 20,
                "marginLeft": 5,
                "maxWidth": 650,
              },
            ]
          }
          testID="button-title"
          textColor="#eb0055"
        >
          Je supprime mon compte
        </Text>
      </View>
      <View
        numberOfSpaces={4}
        style={
          Array [
            Object {
              "height": 16,
            },
          ]
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderRadius": 24,
            "borderWidth": 0,
            "flexDirection": "row",
            "height": 48,
            "justifyContent": "center",
            "maxWidth": 500,
            "opacity": 1,
            "paddingBottom": 2,
            "paddingLeft": 2,
            "paddingRight": 2,
            "paddingTop": 2,
            "width": "100%",
          }
        }
        testID="button-container"
      >
        <Text
          adjustsFontSizeToFit={false}
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontFamily": "Montserrat-Bold",
                "fontSize": 15,
                "lineHeight": 20,
                "marginLeft": 5,
                "maxWidth": 650,
              },
            ]
          }
          testID="button-title"
          textColor="#ffffff"
        >
          Abandonner
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  />
  <View
    customHeight={8}
    style={
      Array [
        Object {
          "height": 8,
        },
      ]
    }
  />
</View>
`;
